### PR TITLE
[discover-2.0] update interfaces and services for query, data, languages

### PR DESCRIPTION
**Pull request by kavilla**
### Description

This PR stabilizes the Multi-Query Language (MQL), and Multiple-Data-Sources (MDS) features by updating types to utilize a generic structure, abstracting data querying, and introducing a Language service. These changes aim to resolve conflicts in data source definitions across different projects and provide a flexible framework for handling various data structures.

## Notes
- The commits on this branch were reviewed individually.
- The size of this PR is primarily due to cleanup and consolidation efforts.

This PR represents a significant step towards a more flexible and extensible multi-query language system in OpenSearch Dashboards.

## Key Changes

1. **Generic Data Structure**: Introduced a flexible `DataStructure` interface that can represent various data hierarchies without hard-coding specific fields.

2. **Abstract Data Querying**: Implemented a system where actual data querying is left to the project or developer to implement, allowing for diverse opinions and implementations.

3. **Language Service**: Added a Language service that registers more closely with the query object, improving the flow of queries through the application.

4. **Consolidation**: Cleaned up and consolidated existing code, resulting in a large PR despite individual commits being reviewed separately.

## DataStructure Interface

The core of this update is the `DataStructure` interface:

```typescript
export interface DataStructure {
  id: string;
  title: string;
  type: string;
  parent?: DataStructure;
  children?: DataStructure[];
  hasNext?: boolean;
  columnHeader?: string;
  meta?: DataStructureMeta;
}
```

This interface allows for representation of various data hierarchies, such as OpenSearch clusters, indices, or even complex nested structures like S3 data sources.

## Benefits

- Flexibility: Developers can populate the DataStructure with values specific to their data source types without modifying the core interface.
- Consistency: Standardized structure for representing different data sources and their hierarchies.
- Extensibility: Easy to add new data source types without changing the underlying system.
Improved Query Handling: The Language service allows for better integration of queries with different data structures.

## Scope
Currently, these changes are only enabled for the Discover application. They lay the groundwork for potential future expansion to other parts of the system.

## Testing the changes

local and CI

## Changelog

- refactor:  Multi-datasources and multi-query languages features to use generic structured types, abstract data querying, and the language service

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by ashwin-pc — APPROVED** [review]

File: src/plugins/data/common/datasets/types.ts
Comment (sejli) [inline src/plugins/data/common/datasets/types.ts:1]: Do we want to mark these as experimental for now? Or are these defined?

File: src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+          : undefined,
+      } as IndexPatternSpec;
+      const temporaryIndexPattern = await this.indexPatterns?.create(spec);
+      if (temporaryIndexPattern) {
+        this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+      }
+    }
+  }
+
+  public fetchOptions(
```
Comment (sejli) [inline src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:85]: TODO: handle caching here?

File: src/plugins/data/public/ui/dataset_selector/dataset_explorer.tsx
Comment (sejli) [inline src/plugins/data/public/ui/dataset_selector/dataset_explorer.tsx:1]: Amazing 💕

File: src/plugins/data/public/ui/dataset_selector/dataset_selector.tsx [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  const togglePopover = useCallback(async () => {
+    if (!isOpen) {
+      await fetchDatasets();
+    }
+    setIsOpen(!isOpen);
+  }, [isOpen, fetchDatasets]);
+
+  const closePopover = useCallback(() => setIsOpen(false), []);
+
+  const options = useMemo(() => {
```
Comment (sejli) [inline src/plugins/data/public/ui/dataset_selector/dataset_selector.tsx:74]: Any reason why we don't keep index patterns as part of the datasets?

**Review by sejli — APPROVED** [review]
Thanks Rocky and Ashwin!

**Review by ashwin-pc — APPROVED** [review]

**Review by sejli — DISMISSED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Comment by AMoo-Miki** [comment]
Manually backported with #7863

**Pull request merged**